functorch/experimental/_cond.py

```diff
@@ -27,6 +27,65 @@
 class UnsupportedAliasMutationException(RuntimeError):
     reason: str
 
+def exported_cond(op, *args):
+    exclude_keys = DispatchKeySet(DispatchKey.FuncTorchDynamicLayerFrontMode) \
+        .add(DispatchKey.FuncTorchDynamicLayerBackMode) \
+        .add(DispatchKey.Functionalize)
+    with _ExcludeDispatchKeyGuard(exclude_keys):
+        with torch.utils._python_dispatch._disable_current_modes():
+            def from_fun(t):
+                if isinstance(t, torch.Tensor):
+                    def to_hint(sym_int) -> int:
+                        if isinstance(sym_int, torch.SymInt):
+                            return sym_int.node._hint
+                        return sym_int
+                    size = pytree.tree_map(to_hint, t.size())
+                    stride = pytree.tree_map(to_hint, t.stride())
+                    return torch.empty_strided(size, stride, requires_grad=t.requires_grad)
+                # Need to specialize symbool for now. Won't affect traced graph.
+                elif isinstance(t, torch.SymBool):
+                    return t.node._hint
+                elif isinstance(t, torch.fx.proxy.Proxy):
+                    raise RuntimeError(f"Unable to symbolically trace HigherOrderOperators {op}")
+                return t
+
+            new_args = pytree.tree_map(from_fun, args)
+
+            # we need to wrap true_fn/false_fn up otherwise the local scope
+            # will contain the original true_fn and false_fn's signature
+            def wrapper(new_args):
+                return cond(*new_args)
+
+            # need to do it together otherwise will need to merge input ->
+            # duplicated effort with what we have already
+            gm, guards, example_inputs = torch._dynamo.export(wrapper, new_args, rewrite_sig=False)
```

Comment:
Things we should figure out:
- the function returned by export accepts flattened inputs and returns flattened outputs. Are we hooking up the args and returns correctly?
- materialization of the example inputs via empty_strided. Ideally export has a mode where it does something reasonable if we pass it FakeTensor inputs.
- caching. Ideally we would cache here: if the user calls cond and the "guards" haven't changed (in quotes because idk if that is the same as the guards object), then we should not need to do a re-trace. Might not need to happen in this PR?

Reply:
These are all good suggestions. A few thoughts right now:
> Are we hooking up the args and returns correctly?

The plan is to 1. check the args of cond node 2. trace where the args are from and find the guards installed. They can only be from inputs/params/buffers of "gm" (otherwise the graph module is incomplete and is not runnable). For inputs, they have guards installed in VariableBuilder. For params/buffers, can directly use the value if they are constant and if they are not constant, they also have guards installed in output_graph.register_attr_or_module. 3. find how guards get resolved and checked (e.g. the L["newargs"].." so that we could do the checking and patching.

> Ideally export has a mode where it does something reasonable if we pass it FakeTensor inputs.

Yeah, worth trying a bit!

> caching.

yes, some python level-caching can be implemented.